ui/pages/editor.tsx

```diff
@@ -1,110 +1,255 @@
-          window.location.href = "/" + window.location.search;
-        } else {
-          res.text().then((msg) => {
-            window.alert(msg);
-          });
-        }
-      })
-      .catch((err) => {
-        console.error(err);
-      });
-  }
-
-  const handleInitOpenFile = (filePath:string) => {
-    // TODO - highlight/expand proper file in filetree
+    saveStaticFile(fileName, valueGetter.current(), () => {
+      mutate("/static/" + fileName, valueGetter.current());
+      editorModels[fileName].saveModel();
+    }, (err, msg) => {
+      if(err) console.log(err);
+      if(msg) window.alert(msg);
+    });
   }
 
+  // TODO: bug here if you switch too fast
   const fileOpenCallback = (filePath: string): (() => void) => {
     return () => {
-      setName(filePath);
+      selectFile(filePath);
     };
   };
 
-  const _transformFileTreeData = (fileTreeData): IFileTreeNode[] => {
-    let fileIdRef = { id: -1 };
-    const transformFileDataHelper = (fileTreeData, fileIdRef, pathParts) => {
+  // Turn server filetree data into data that can be read by the FileTree/Blueprint (TODO: possibly unify these)
+  const initializeFileTreeData = (fileTreeData): void => {
+    const filesMap = {};
+    const transformFileDataHelper = (fileTreeData, pathParts) => {
       const data: IFileTreeNode[] = [];
       for (const levelName in fileTreeData) {
-        fileIdRef.id += 1;
         pathParts.push(levelName);
         const { _type, ...children } = fileTreeData[levelName];
         if (_type === "file") {
-          data.push(new FileNode(fileIdRef.id, levelName, fileOpenCallback(path.join(...pathParts))));
+          const fileKey = path.join(...pathParts);
+          const node = new FileNode(uuidv4(), levelName, fileOpenCallback(fileKey));
+          data.push(node);
+          filesMap[fileKey] = node;
         } else {
           data.push(
-            new DirectoryNode(fileIdRef.id, levelName, transformFileDataHelper(children, fileIdRef, pathParts)),
+            new DirectoryNode(uuidv4(), levelName, transformFileDataHelper(children, pathParts)),
           );
         }
         pathParts.pop();
       }
 
       return data;
     };
-
-    return transformFileDataHelper(fileTreeData, fileIdRef, []);
+    
+    const treeData = transformFileDataHelper(fileTreeData, []);
+    setFileTree(treeData);
+    setFileNodeMap(filesMap);
   };
 
-  const language = extToLanguage(name.split(".")[1]);
+  const saveStaticFile = (fileName, content, onSuccess, onError) => {
+    const url = "/static/" + fileName + window.location.search;
+    fetch(url, {
+      method: "PUT",
+      body: content,
+    }).then((res) => {
+      if (res.status === 200) {
+        onSuccess(res);   
+      } else {
+        res.text().then((msg) => {
+          const err = new Error(`Unexpected status: ${res.status}`);
+          onError(err, msg);
+        });
+      }
+    })
+    .catch((err) => {
+      onError(err);
+    });
+  }
```

Comment:
Nit: Async/Await instead of Promise chains

Reply:
Lets make that the standard